[9.0] MIG bi_sql_editor

nicomacr opened the pull request.

legalsylvain committed:
new module bi_sql_editor [IMP] function to guess model for many2one field; [ADD] security

legalsylvain committed:
typo

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

nicomacr committed:
[MIG][9.0]bi_sql_editor

nicomacr committed:
ADD oca_dependencies

nicomacr committed:
DEL old files

AaronHForgeFlow commented:
I :heart: this module. Would you include the changes proposed here https://github.com/OCA/reporting-engine/pull/149 ?

JordiBForgeFlow commented:
@aheficent perhaps you can propose a PR that includes it

legalsylvain commented:
please wait, i'm doing some improvement on #149.

legalsylvain committed:
[ADD] create ui now create tree view
[IMP] fr translation
[IMP] add last refresh date time on the action name of each sql materialized view

legalsylvain commented:
improvment available here : https://github.com/OCA/reporting-engine/pull/150

legalsylvain commented:
Hi @aheficent. Thanks a lot for porting this module. LGTM for me (code / test). please ping me once you integrated #150 commit.
regards.

AaronHForgeFlow commented:
Sure @legalsylvain Thank you for the improvement

JordiBForgeFlow commented:
@nicomacr @legalsylvain @aheficent I have improved 9.0 with the cherry pick of #150 here: https://github.com/adhoc-dev/reporting-engine-1/pull/1

nicomacr commented:
@jbeficent thanks for the PR

AaronHForgeFlow committed:
[UPT]add test. FIx unlink

legalsylvain commented on bi_sql_editor/demo/bi_sql_view.xml:
```diff
@@ -0,0 +1,59 @@
+SELECT
+    name as x_name,
+    case
+        when author ilike '%OpenERP SA%' THEN 'Odoo SA'
```
because, odoo renamed authorship in V9.

    case
        when author ilike '%OpenERP SA%' THEN 'Odoo SA'
        when author ilike '%Odoo SA%' THEN 'Odoo SA'
        when author ilike '%Odoo Community Association (OCA)%' THEN 'OCA'
       else 'Undefined Author' END as x_author_type

AaronHForgeFlow commented:
Hi @nicomacr I'm proposing some tests here: https://github.com/adhoc-dev/reporting-engine-1/pull/2 . Also it fixes the unlink method, which is not calling super.

legalsylvain reviewed (approved):
LGTM to me. one comment on demo data to update.

Otherwise, :+1: 
Thanks ! 

PS : could you review, the V8.0 PR ? 
https://github.com/OCA/reporting-engine/pull/150

AaronHForgeFlow commented:
I closed the previous PR for the test. I created a new one here: https://github.com/adhoc-dev/reporting-engine-1/pull/3 but it's failing because it's creating a lot of connections to the database. I'll update the PR when this issue is solved

legalsylvain commented:
Any other review to validate the PR ? thanks !

JordiBForgeFlow commented:
Can we add an option for the person thay creates the query to decide the order of the.views? (Treen graph, pivot)?

AaronHForgeFlow commented:
@legalsylvain This works ok to me. However there's must be something wrong with the test I did because when I run it on local it's trying to connect to the database multiple times. Getting this:
PoolError: The Connection Pool Is Full

Honestly, don't know the reason

elicoidal commented on bi_sql_editor/README.rst:
```diff
@@ -0,0 +1,184 @@
+  take a while, if view is materialized)
+
+* If it's a MATERIALIZED view: 
+    * a cron task is created to refresh
```
add a blank line

elicoidal commented on bi_sql_editor/README.rst:
```diff
@@ -0,0 +1,184 @@
+
+To use this module, you need to:
+
+* Go to 'Reporting' / 'Custom Reports' 
```
Best would be enumerated list (nice to have)

elicoidal commented on bi_sql_editor/__openerp__.py:
```diff
@@ -0,0 +1,28 @@
+        'views/menu.xml',
+    ],
+    'demo': [
+        'demo/res_groups.xml',
```
demo/res_groups_demo.xml (several)

elicoidal commented on bi_sql_editor/demo/res_groups.xml:
```diff
@@ -0,0 +1,18 @@
+License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+-->
+
+<openerp><data>
```
odoo
+Improve the format

elicoidal commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -0,0 +1,542 @@
+
+    technical_name = fields.Char(
+        string='Technical Name', required=True,
+        help="Suffix of the SQL view. (SQL full name will be computed and"
```
remove (

elicoidal commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -0,0 +1,542 @@
+    technical_name = fields.Char(
+        string='Technical Name', required=True,
+        help="Suffix of the SQL view. (SQL full name will be computed and"
+        " prefixed by 'x_bi_sql_view_'. Should have correct"
```
Syntax should follow: https://www.postgresql.org/"
        "docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

elicoidal commented on bi_sql_editor/models/bi_sql_view_field.py:
```diff
@@ -0,0 +1,225 @@
+
+    ttype = fields.Selection(
+        string='Field Type', selection=_TTYPE_SELECTION, help="Type of the"
+        " Odoo field that will be created. Let empty if you don't want to"
```
s/let/keep

elicoidal commented on bi_sql_editor/models/bi_sql_view_field.py:
```diff
@@ -0,0 +1,225 @@
+    many2one_model_id = fields.Many2one(
+        comodel_name='ir.model', string='Model',
+        help="For 'Many2one' Odoo field.\n"
+        " Co Model of the field.")
```
Comodel

elicoidal reviewed (changes requested):
nice one

legalsylvain commented:
> Can we add an option for the person that creates the query to decide the order of the.views? (Treen graph, pivot)?

We could 
- add an extra char field in the view, with a default value to 'graph,pivot,tree', (readonly if sql is not in the draft / analyzed state)
- add a constraint that check that value.split(',') contains only ['graph', 'pivot', 'tree']
- use this value to generate the action. (and use it in the button_open_ui)

No time for this feature, for my side, but ok to review it. (It could be done in the second PR.)

AaronHForgeFlow commented:
@legalsylvain I'll propose a PR for that

AaronHForgeFlow committed:
[IMP]add an option to select the first view

AaronHForgeFlow commented:
Hi @nicomacr @legalsylvain I added an option for the person that creates the query to decide the first view to show: https://github.com/adhoc-dev/reporting-engine-1/pull/4

jjscarafia committed:
Merge pull request #4 from Eficent/9.0-mig-bi_sql_editor-imp

[IMP]bi_sql_editor. Select View Order.

LoisRForgeFlow reviewed (approved):
I agree with @elicoidal remarks but apart from that this module is great!

Thank you for the great contribution.

legalsylvain commented:
@nicomacr , @aheficent Could you make minor fixes asked by @elicoidal  ? 

:+1: Thanks a lot again. (The feature to select view types is a great improvement !)

regards.

JordiBForgeFlow commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -0,0 +1,568 @@
+
+    _STATE_SQL_EDITOR = [
+        ('model_valid', 'SQL View and Model Created'),
+        ('ui_valid', 'Graph, action and Menu Created'),
```
Please rename `Graph, action and Menu Created` to `Views, Action and Menu Created`, because now you show pivot and tree as well.

AaronHForgeFlow committed:
[FIX]Review round

AaronHForgeFlow commented:
Hi @legalsylvain @nicomacr I just created a new Pull Request for the fixes proposed here. Can you review? Thanks https://github.com/adhoc-dev/reporting-engine-1/pull/5

legalsylvain commented:
2 bugs on CI.
1. ir.model.fields.create() includes unknown fields: tree_visibility
2. No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself. -> something wrong happened. don't know why.

JordiBForgeFlow commented:
@pedrobaeza @gurneyalex do you know why would Travis stall on this one?

AaronHForgeFlow commented:
@jbeficent @legalsylvain I think it's something related the tests. When I run those on local I got the following: `PoolError: The Connection Pool Is Full`

AaronHForgeFlow commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -0,0 +1,568 @@
+            sql_view.rule_id.unlink()
+            if sql_view.cron_id:
+                sql_view.cron_id.unlink()
+            sql_view.write({'state': 'draft', 'has_group_changed': False})
```
It seems to me that the bi_sql records created in a test are never deleted. And when the bi_Sql record is validaded never goes back to draft. Besides, this line is giving errors when calling this method from a test. 

@astirpe Maybe you can shed some light here

BTW: An uninstall hook is needed to drop some materialized views and some relations

astirpe replied:
Strange error. I can take a look at it next week.

astirpe replied:
Also some warnings are raised:
```
2017-07-25 16:16:59,706 172 WARNING openerp_test openerp.models: ir.model.fields.create() includes unknown fields: tree_visibility
2017-07-25 16:16:59,710 172 WARNING openerp_test openerp.models: ir.model.fields.create() includes unknown fields: tree_visibility
```

AaronHForgeFlow replied:
This is what happens when running the test several times. The records are never deleted. It seems I need to run some queries every time I run the tests:
![image](https://user-images.githubusercontent.com/19620251/28662621-10cfb594-72bb-11e7-9c10-8f481c5858cd.png)

Moreover this have happened after running the uninstall hook, which looks like:
` ```
   cr.execute("""
        select relname
        from pg_class
        where relname like 'x_bi_sql_view%' and relkind='m'
        """)
    for r in cr.fetchall():
        cr.execute("""
               DROP MATERIALIZED VIEW %s
            """ % r)
```

    cr.execute("""
        select relname
        from pg_class
        where relname like 'x_bi_sql_view%' and relkind='r'
        """)
    for r in cr.fetchall():
        cr.execute("""
               DROP TABLE %s
            """ % r)
    cr.execute("""
        select table_name from INFORMATION_SCHEMA.views
        where table_name like 'x_bi_sql%'""")

    # Drop not materialized views
    for v in cr.fetchall():
        cr.execute("""
               DROP VIEW %s
            """ % v)

    #Drop table if uninstalling went wrong
    cr.execute("""
        delete from ir_model_fields where model like 'bi.sql.view%';
        delete from ir_model_fields where model like 'bi_sql_%';
        delete from ir_model where model like 'x_bi_sql_view.%';
        """)`

AaronHForgeFlow replied:
I'll propose another Pull Request fixing this

AaronHForgeFlow commented on bi_sql_editor/demo/bi_sql_view_demo.xml:
```diff
@@ -0,0 +1,57 @@
+        </field>
+    </record>
+
+    <function model="bi.sql.view" name="button_validate_sql_expression" eval="([ref('module_sql_view')])"/>
```
This is causing an error when uninstalling and installing the module:
`relation "x_bi_sql_view_modules_view" already exists`

I'm trying to do an uninstalling hook including this and some other clauses to delete remaining tables and views.

astirpe replied:
Looks like the same problem that was occurring when uninstalling and installing the bi_view_editor: an hook script like [this one](https://github.com/OCA/reporting-engine/blob/9.0/bi_view_editor/hooks.py) would solve this issue

AaronHForgeFlow committed:
[IMP]add uninstall hook. Add checks when deleting records.

AaronHForgeFlow commented:
I proposed a new Pull Request here: https://github.com/adhoc-dev/reporting-engine-1/pull/6

@nicomacr Please take a look at this before it's merged.

To sum up, the Pull request does the following:

- Adds uninstall hook: Avoid errors when reinstalling.
- Adds checks when deleting records in the button_set_draft method
- For non materialized views was getting errors when deleting the views. Added check
- Deleted test. No matter what I tried. Records created on the test are never deleted and consequently the views, the tables and so on. Moreover, if the test refers to a record created with demo data, the system enters in a loop, and start creating connections to the database to create views, tables and so on. Insane. I tried to reproduced manually and this was not happening, so I think it is a problem with the test environment. To sum up, test are deleted.

legalsylvain commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -208,13 +208,20 @@ def button_set_draft(self):
-            sql_view.action_id.unlink()
-            sql_view.menu_id.unlink()
-            sql_view.rule_id.unlink()
+            if sql_view.tree_view_id:
```
why did you added a test ? if rule_id is False, sql_view.r_le_id.unlink() make nothing.

legalsylvain commented on bi_sql_editor/hooks.py:
```diff
@@ -0,0 +1,52 @@
+        rec.button_set_draft()
+        rec.unlink()
+
+    # delete dirty data that could cause problems
```
I don't understand all the lines below. 15+
for me button_set_draft() and unlink() are enought, to drop all the (materialized or not) views.

AaronHForgeFlow replied:
I'll check that again without tests. But with them that was not enough, because it creates a lot of dirty data in the database. Thank you @legalsylvain

legalsylvain commented on bi_sql_editor/hooks.py:
```diff
@@ -0,0 +1,52 @@
+
+    # Drop table if uninstalling went wrong
+    cr.execute("""
+        delete from ir_model_fields where model like 'bi.sql.view%';
```
better to do sql_view.model_id.unlink(), don't you think ?

AaronHForgeFlow replied:
I think the same. This is supposed to be done by this line `rec.button_set_draft()` But for some reason it does not work properly. At the end it shows this error:

> Unable to delete 121@ir.model
and then:
> UserError: (u"The field 'bi.sql.view.field.bi_sql_view_id' cannot be removed because the field 'bi.sql.view.bi_sql_view_field_ids' depends on it.", None)

That's why I put all those queries.

legalsylvain commented on bi_sql_editor/models/bi_sql_view.py:
```diff
@@ -464,7 +476,8 @@ def _drop_model_access(self):
     def _drop_model_and_fields(self):
         for sql_view in self:
-            sql_view.model_id.unlink()
+            if sql_view.model_id:
```
same here.

legalsylvain reviewed (changes requested):
Hi thanks a lot for adding a hook. 
Some comments inline however.

thanks.

AaronHForgeFlow commented:
BTW, travis compilation was not failing because of the test here but for a test in another module: test_report_qweb_pdf_watermark

MiquelRForgeFlow commented:
No, Travis now is not failing by the pdf_watermark. It seems it is failing because it cannot install some external libraries. They should be added in Travis or something like this may be.

AaronHForgeFlow commented:
@mreficent All right. I'll try to fix that

MiquelRForgeFlow commented:
@aheficent sorry, yes, the problem is with `report_qweb_pdf_watermark`

AaronHForgeFlow commented:
BTW among other things, Lint check fails because of the queries in the uninstall hook. But those queries are necessary IMO so I'd keep that

AaronHForgeFlow committed:
[FIX]lint check. unnecessary queries in uninstall hook.

AaronHForgeFlow commented:
Hi @legalsylvain I created another Pull Request here to attend your comments. Despite my previous comment I realized that those queries in the uninstall hook are not needed. However this error is still shown when uninstalling:
> Unable to delete 121@ir.model
and then:
> UserError: (u"The field 'bi.sql.view.field.bi_sql_view_id' cannot be removed because the field 'bi.sql.view.bi_sql_view_field_ids' depends on it.", None)

Nevertheless, no bi.sql.view and no bi.sql.view.field remains in the database when uninstalling. That's why I removed the queries from the uninstall hook.

AaronHForgeFlow commented:
Created Pull Request here to fix flake8 error https://github.com/adhoc-dev/reporting-engine-1/pull/8

However this is still pending:
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection

Does anybody knows how to configure travis file to avoid this check? :pray:

legalsylvain commented:
@aheficent : 

``self.env.cr.commit() # pylint: disable=invalid-commit``

Sample : https://github.com/OCA/server-tools/blob/10.0/database_cleanup/models/purge_columns.py#L51

regards;

AaronHForgeFlow commented:
Thank you @legalsylvain! I added this to my last Pull Request. (https://github.com/adhoc-dev/reporting-engine-1/pull/8)

AaronHForgeFlow committed:
[FIX]pylint

legalsylvain commented:
@jbeficent , @aheficent could you set a little +1 on the PR against V8 you cherry picked ? 
https://github.com/OCA/reporting-engine/pull/150.

thanks !

AaronHForgeFlow commented:
Runbot is compiled. It's showing red because of this:
`ParseError: "no such savepoint
" while parsing /home/odoo/build/OCA/reporting-engine/bi_sql_editor/demo/bi_sql_view_demo.xml:52, near
<function model="bi.sql.view" name="button_validate_sql_expression" eval="([ref('module_sql_view')])"/>`

Should we create a savepoint before calling the method  button_validate_sql_expression?

Edit: Run in local in debug mode. Same error as here travis is getting. In the second try it fails for different reason, the relations created for the demo data in the previous try were still there. Same problem as the one I was having with the tests.

AaronHForgeFlow committed:
[FIX]remove  cr.commit, that removed the savepoint
 [UPT]add test

AaronHForgeFlow commented:
Hi @nicomacr I created a new PR here: https://github.com/adhoc-dev/reporting-engine-1/pull/9

This fixes the error related with the save point that was introduced in the last commit. Besides I added some tests back. However, the tests don't cover the part of creating the models because that would leave unwanted data in the database.
@legalsylvain @jbeficent

AaronHForgeFlow committed:
[FIX]lint errors

AaronHForgeFlow commented:
@nicomacr Sorry for the Lint errors. :face_with_head_bandage:  Created new PR for that:https://github.com/adhoc-dev/reporting-engine-1/pull/10   Probably a commit squashing will be needed later

AaronHForgeFlow committed:
[FIX]ir_model does not have field tree visibility

AaronHForgeFlow commented:
Created PR to prevent runbot warning https://github.com/adhoc-dev/reporting-engine-1/pull/11

Travis will be fixed when this is merged: https://github.com/OCA/reporting-engine/pull/154

legalsylvain commented on bi_sql_editor/models/bi_sql_view_field.py:
```diff
@@ -174,8 +174,6 @@ def _prepare_model_field(self):
             'relation': self.ttype == 'many2one' and
             self.many2one_model_id.model or False,
-            'tree_visibility': self.field_description and
-            'available' or 'unavailable',
```
thanks !

JordiBForgeFlow reviewed (approved):
Technical and functional review :+1:

JordiBForgeFlow commented:
Travis is KO due to problem found in #154. But tested locally, and all test are passing correctly.

The only problem we found is that it is not possible to include in the tests the parts where the models are created, because every time that you pass the test the tables created by the test are not deleted from the DB -- very strange --, so the next time you run the test, it fails because the table already exists.

legalsylvain commented:
Well, I propose to merge this PR, because it's OK. (and it has 6 :+1:, that's a record ! )
let's go improving test in another PR, or for another version. (8.0 version didn't include test at all, so, it's better).
